README.md
Curecode
========

Lightweight web-based bugtracker 

private/ChangeLog.txt
CureCode "R3 Edition" ChangeLog :

Changelog : 0.9.13 - ??/??/2011

	o FEAT: Search field now accepts multiple keywords.
	

Changelog : 0.9.12 - 13/12/2010

	o FEAT: Added the ability to move tickets between projects.
	o FEAT: New search field: "User".
	o FEAT: New isolated search field: "Commented by".
	o FEAT: RSS feed added for projects & tickets changes.
	o FEAT: Short URLs for tickets direct referencing added.
	
	o FIX: double line breaks in PRE tags in comments removed.
	o FIX: stats bar graphs were never displaying per-project stats.
	o FIX: double escaping of HTML entities in description and comments removed.
	

Changelog : 0.9.11 - 20/02/2010

	o FEAT: Custom captcha system replaced by ReCaptcha.
	o FEAT: Auto-adjusting scale for bar charts in stats (more accurate and readable).
	o FEAT: Category selector added in tickets list detailed view
	o FEAT: Empty attached file section not shown anymore in tickets read-only view.
	
	o FIX: When clicked, attached files will now show up in browser's window by default.
	o FIX: When attaching file in a new ticket, the returned attached filename was 'none'.
	

Changelog : 0.9.10 - 17/01/2010

	o FEAT: Ticket's attached files management added.
	o FEAT: Emails are not sent using Cheyenne's built-in MTA.
	
	o FIX: Default query parameters corruption for anonymous users fixed.
	o FIX: Issue #1371 regarding policy for ticket deletion by owner.
	o FIX: Dimissed ticket can't be changed by the owner anymore.
	o FIX: Ordering not working properly for all filters and columns.
	o FIX: Minor missing translations for french catalog.

	o LOOK: Added Henrik's 33% fix for page header.
	o LOOK: "Example code" field now use PRE tag for better rendering of source code.


Changelog : 0.9.9 - 26/07/2009

	o FEAT: Prefered filters and project can now be user-defined.
	o FEAT: Categories order can now be set by admin.
	
	o FIX:  Stats charts are now correct.
	o FIX:  Default filter in anonymous mode works now correctly.
	o FIX:  Switched from HTML quirk mode to correct doctype.
	
	o LOOK: Fixed borders issues in CSS (except for IE8)


Changelog : 0.9.8 - 26/05/2009

	o FEAT: "User" column added to tickets list page.
	o FEAT: Search text feature extended to comments.
	o FEAT: "Main" page now populated with statistics.
	o FEAT: New Preset filter: "My Reports" (default when logged)
	
	o FIX:  "By Submitter" filter now working correctly.
	o FIX:  Redirection issues with Chrome fixed.
	o FIX:  In Preset mode, columns sorting links removed.
	
	o LOOK: Textarea edit fields now enlarged to maximum width.
	

Changelog : 0.9.7 - 11/04/2009

	o FEAT: Dismissed tickets are now filtered out from "Active Priorities" and "Worst Severity" filters
	o FEAT: Current project name added to "Add Ticket" form.
	o FEAT: Reversed sort order for Versions and Categories lists in admin page.
	o FEAT: Added an experimental API for getting tickets in REBOL format.

	o FIX:	Front page stats now count "dismissed" tickets as closed.
	o FIX:	Message "n ticket(s) found" fixed for localization, now translates correctly in french.
	

Changelog : 0.9.6 - 21/01/2009

	o FEAT: Ticket submitter can now delete his ticket.
	o FEAT: Admin can now delete any ticket.
	o FEAT: Admin can now delete a user.
	o FEAT: Admin can now change user's role.
	o FEAT: Admin can now force validation of a new user's account.
	
	o FIX:	Tickets list loading twice on first visit issue fixed.
	
	o LOOK: Input fields and buttons border width reduced to 1 pixel.


Changelog : 0.9.5 - 05/12/2008

	o FEAT: Added Preset ticket list filters (from Altme Tracker)
	o FEAT: Added search ticket by text feature
	o FEAT: Added direct access to a ticket knowing its ID
	o FEAT: Filter zone in tickets list page redesigned to account for different filtering modes.
	o FEAT: Back and forward navigation now possible from inside a ticket.
	o FEAT: Ticket's ID in descriptions and comments now displayed as links.
	
	o FIX:  Parsing URL in description/comments section will now stop on #"<".
	o FIX:  CSS fixed for tables border rendering on non-IE browser.
	o FIX:  Case-only changes in ticket's title are now possible.
	
	o LOOK: Minor positionning enhancements of login block in home page.
	o LOOK: "Description" and "Code example" sections default height reduced.
	
R3 Tracker Migration :

	o "Response" AltMe Tracker field now map to a comment in CC (keeping the original response's date).

	
	

private/generate-test.r
REBOL []

random/seed now/time/precise 

if not find first system/schemes 'MySQL [
	do %mysql-protocol.r
]

make-text: has [out][
	clear out: ""
	loop 32 [append out #"a" + ((random 26) - 1)]
	out
]
prin "Database root "
db: open mysql://root:?@localhost/bugs


send-sql db rejoin [ {
	INSERT INTO users VALUES (
		NULL,
		'dk1',
		'098F6BCD4621D373CADE4E832627B4F6',
		'default@curecode.org',
		'00000000000000000000000000000000',
		NOW(),
		2,
		NOW()
	)}
]
send-sql db rejoin [
	"INSERT INTO projects VALUES (NULL,'Test','This is a test project',0,0)"
]
repeat ver 4 [
	send-sql db rejoin [
		"INSERT INTO versions VALUES (NULL,2,'1." ver "',0)"
	]
]
send-sql db rejoin ["INSERT INTO categories VALUES (NULL,2,'User Interface',0)"]
send-sql db rejoin ["INSERT INTO categories VALUES (NULL,2,'Security',0)"]
send-sql db rejoin ["INSERT INTO categories VALUES (NULL,2,'Text',0)"]
send-sql db rejoin ["INSERT INTO categories VALUES (NULL,2,'Performances',0)"]


counts: [0 0 0 0]

repeat cnt 10000 [
	either zero? cnt // 1000 [prin newline prin cnt][
		if zero? cnt // 100 [prin #"."]
	]
	poke counts 2 counts/2 + 1
	send-sql db rejoin [ 
		"INSERT INTO tickets VALUES (NULL,"
		2 ","
		"'" join make-text "',"
		"'" join make-text "',"
		"'" join make-text "',"
		1 + random 4 ","
		1 + random 4 ","
		1 + random 4 ","
		random 8 ","
		random 7 ","
		random 9 ","
		random 6 ","
		random 6 ","
		"0,"
		2 ","
		"NOW(), NOW())"
	]
]

repeat cnt 4 [
	send-sql db rejoin ["UPDATE projects SET count=" counts/2 " WHERE id=" 2]
]

close db

halt

app-init.r
REBOL [
	Purpose: "RSP environement init code"
]

on-application-start: does [
	do %private/helper.r
	do %private/db-abstract.r
	do %private/access-control.r
]

on-application-end: does []

on-database-init: does [
	define-cached-queries
]

on-session-start: has [list][
	session/add 'project-id 1
	session/add 'user-id 	none
	session/add 'user-name	none
	session/add 'user-role	'viewer
	session/add 'user-prj-role none
	session/add 'prefs	none
	session/add 'files	none
	session/add 'query copy [
		preset		"search"
		filter		2
		text		#[none]
		fields		1
		tid			#[none]
		type		0
		severity	0
		user		0
		status		0
		hstatus		0
		priority	0
		results		25
		action		search
		page		1
		orderby 	1
		version		#[none]
		fixedin		#[none]
		cat			#[none]
		resolv		#[none]
		find		#[none]
		apply		#[none]
	]
]

on-session-end: does []

on-page-start: does [
	filter-access
]

on-page-end: does []


private/helper.r
REBOL []

random/seed now/time/precise 

zeropad: func [value n][
	head insert/dup value: form value #"0" n - length? value
]

ceil: func [n [number!]][
	if integer? n [return n]
	to integer! n + 0.99999999999999
]

mark: context [
	selected: func [key field [word!] default][
		if key = any [request/content/:field :default][prin " selected "]
	]
	
	error: func [field [word!]][
		if find invalid :field [prin {class="error"}]
	]
]

encode-pass: func [pass [string!]][
	enbase/base checksum/method pass 'md5 16
]

pre-chars: charset "^/^-"
pre-encode: func [data [string!] /local s][
	data: convert data [pre-chars][
		either value = "^/" [
			"<br/>"
		][
			"&nbsp;&nbsp;&nbsp;&nbsp;"
		]
	]
	parse/all data [				;-- remove extra <br/> encodings only
		any [
			"<pre" any [
				s: <br/> (s: remove/part s 5) :s
				| </pre> break
				| skip
			]
			| skip
		]
	]
	data
]

digits: charset "0123456789"
not-ws: complement charset "^/^-^M <"
activate-links: func [data [string!] /local s e link][
	parse/all data [
		any [
			"&#"
			| s: #"#" some digits e: (
				change/part s link: rejoin [
					{<a href="ticket.rsp?id=} next link: copy/part s e {">}
					link 
					"</a>"
				] e
				s: skip s length? link
			) :s
			| s: "http://" some not-ws e: (
				change/part s link: rejoin [
					{<a href="} link: copy/part s e {">}
					link 
					"</a>"
				] e
				s: skip s length? link
			) :s
			| skip
		]
	]
	data
]

coltitle: func [pos show? name /local current?][
	if not show? [
		prin say name
		exit
	]
	if current?: pos = abs order [
		prin {<img src="img/} 
		prin pick ["up" "down"] positive? order
		prin {.png" align="right">}
	]
	prin {<a href="view-tickets.rsp?orderby=}
	prin any [all [current? negate order] pos]
	prin {">}
	prin say name
	prin {</a>}
]

utf8-encode: func [
	"Encodes the string data to UTF-8 (from Latin-1)"
	str [any-string!] "string to encode"
	/local c h
][
	;if you remove 'copy you can change the original string
	parse/all copy str [
		any [
			h: skip ( if 127 < c: first h [
				h: change h c / 64 or 192
				insert h c and 63 or 128
			]) :h
		   skip
		]
	]
	head h
]

quotes-chars: charset {'^"}
js-encode: func [str [any-string!]][
	convert utf8-encode str [quotes-chars][head insert value #"\"]
]

export-json: func [/crunch] [
	response/set-header 'Content-Type "application/json"
]

to-epoch: func [dt [date!]][
	(dt/date - 01/01/1970) * 86400 + to integer! dt/time
]

chart-form: func [blk sep /with tot /local out][
	out: make string! 3 * length? blk
	
	foreach v blk [
		either tot [
			insert tail out to integer! 100 * either zero? tot [0][divide to integer! v tot]
		][
			insert tail out utf8-encode say v
		]
		insert tail out sep
	]
	head remove back tail out
]

chart-form-full: func [data colors labels tot /local outd outc outl i list][
	outd: make string! 3 * length? data
	outc: make string! 64
	outl: make string! 64
	
	i: 1
	foreach v data [
		unless zero? v: to integer! v [		
			insert tail outd to integer! 100 * either zero? tot [0][v / tot]
			insert tail outd #","
			
			insert tail outc colors/:i
			
			insert tail outl utf8-encode say labels/:i
			insert tail outl #"|"
		]
		i: i + 1
	]
	foreach buf list: [outd outc outl][remove back tail get buf]
	reduce list
]


adjust-scale: func [out [block!] max-val [integer!] /local f s val][
	f: pick [50 5] 50 <= (s: max-val / 5)							;-- stepping factor
	s: f * (1 + to integer! s / f)								 	;-- axis steps rounded to ceil
	unless zero? val: max-val // s [max-val: max-val - val + s]		;-- max value adjustment
	insert tail out max-val
	insert tail out s
]

prepare-stockpile: func [list [block!] /local out dt val s][
	sort/reverse/skip list/1 2
	sort/reverse/skip list/2 2
	out: array/initial 3 []
	s: 0
	
	repeat c 6 [
		dt: now/date
		dt/month: dt/month + 1 - c
		dt/day: 1
		insert out/1 join copy/part pick system/locale/months dt/month 3 [#"-" skip form dt/year 2]
		insert out/2 val: any [select list/1 dt "0"]
		if s < val: to integer! val [s: val]
		insert out/3 val: any [select list/2 dt "0"]
		if s < val: to integer! val [s: val]
	]
	adjust-scale out s
	out
]

prepare-monthly-stockpile: has [out dt val s][
	out: array/initial 2 []
	s: 0
	
	repeat c 6 [
		dt: now/date
		dt/month: dt/month + 1 - c
		dt/day: 1
		insert out/1 join copy/part pick system/locale/months dt/month 3 [#"-" skip form dt/year 2]
		dt/month: dt/month + 1
		dt/day: dt/day - 1
		insert out/2 val: get-stockpile-month dt
		if s < val: to integer! val/1 [s: val]
	]
	adjust-scale out s
	out
]

history: context [
	labels: make hash! 57
	foreach [w c][
		added		-
		removed		-
		modified	-
		summary		-
		description	-
		code		-
		severity	severities
		status		statuses
		resolution	resolv
		priority	priorities
		reproduce	reproduces
		version		-
		fixedin		-
		category	-
		comment		-
		ticket		-
		type		types
		platform	platforms
		moved		-
	][
		repend labels [w c uppercase/part form w 1]
	]
	poke labels 11 * 3 "Reproducibility"
	
	encode: func [w [word!]][
		divide 2 + index? find/skip labels w 3 3
	]
	decode: func [fid [integer!] /value vid /local w][
		say either all [
			value
			'- <> w: pick labels 3 * fid - 1 
		][
			select do-sql/flat 'bugs w vid
		][
			pick labels 3 * fid
		]
	]	
]


set 'form-time func [time [time!]][
	copy/part either time/hour < 10 [
		head insert mold time #"0"
	][
		mold time
	] 5
]

set 'to-UTC func [date [date!]][date - date/zone]

make-vkey: has [key][
	key: make string! 16
	loop 16 [append key #"@" + random 26]
	key: checksum/method key 'md5
	enbase/base key 16
]

filters-list: [
	1	"My Reports"
	2	"Most Recent Reports"
	3	"Unreviewed Submissions"
	4	"Active Priorities"
	5	"Tested"
	6	"Recent Changes"
	7	"Worst Severity"
	8	"By Submitter"
	9	"Completed"
]

enum-filters: has [out][
	either session/content/user-id [
		filters-list
	][
		skip filters-list 2
	]
]

apply-filter: func [spec [block!] /local filter][
	filter: pick [
		[orderby 8 user 1]					;-- My Tickets
		[orderby  1]						;-- Most Recent Reports
		[status  1 orderby  4]				;-- Unreviewed Submissions
		[hstatus [6 8 9 10] orderby  6]		;-- Active Priorities
		[status  9 orderby  8]				;-- Tested
		[orderby 8]							;-- Recent Changes
		[hstatus [6 8 9 10] orderby  4]		;-- Worst Severity
		[orderby -10]							;-- By Submitter
		[status  10 orderby -1]				;-- Completed
	] spec/filter
	
	foreach name [type severity status hstatus priority user orderby][
		poke find spec name 2 any [select filter name 0]
	]
]

recaptcha-verify: func [spec [block!] /local data res][
	data: rejoin [
		"privatekey=" 		recaptcha/private-key
		"&amp;remoteip="	request/client-ip
		"&amp;challenge="	spec/recaptcha_challenge_field
		"&&amp;response="	spec/recaptcha_response_field
	]
	res: attempt [
		read/custom http://api-verify.recaptcha.net/verify reduce ['POST data]
	]
	to logic! all [res find/part res "true" 4]
]

get-instance-db: func [instance [string!] /local tests rule list url][
	tests: [
		url: select list 'bugs
		url: find/last url slash
		not empty? next url
		return head url
	]
	all [
		list: select cheyenne-conf/globals 'databases
		all tests
	]
	parse cheyenne-conf rule: [
		any [
			'virtual-root set name string!
			thru 'databases set list block! (
				all [
					list
					instance = trim/with copy name slash
					all tests
				]
			) | into rule | skip
		]
	]
	none
]

; ====== Email management ======

curecode-emitter: no-reply@curecode.org

send-confirmation: func [
	spec [block!] vkey [string!]
	/local url login pass template
][
	url: rejoin [
		request/headers/Host
		either request/server-port = 80 [""][join ":" request/server-port]
		request/web-app
		"/validate.rsp?id=" url-encode spec/login 
		"&key=" vkey
	]
	login: spec/login
	pass: spec/pass
	template: read join locale/get-path %email-activation.tpl
	replace template "$url" url
	replace template "$login" login
	replace template "$pass" pass
	
	send-email compose [
		from: 	 (curecode-emitter)
		to:   	 (to-email spec/email)
		subject: (rejoin ["[" request/config/locals/name "] " say "Account activation"])
	] template
]

send-reset-URL: func [
	spec [block!] vkey [string!]
	/local url template
][
	url: rejoin [
		request/headers/Host
		either request/server-port = 80 [""][join ":" request/server-port]
		request/web-app
		"/reset.rsp?id=" url-encode spec/login 
		"&key=" vkey
	]
	template: read join locale/get-path %email-pass-reset.tpl
	replace template "$url" url
	
	send-email compose [
		from: 	 (curecode-emitter)
		to:   	 (to-email spec/email)
		subject: (rejoin ["[" request/config/locals/name "] " say "Password reset confirmation"])
	] template
]


private/install.r
REBOL [
	Title: "CureCode Installation Script"
	Author: "Nenad Rakocevic/SOFTINNOV"
	Date: 27/01/2010
	Version: 1.0
]

print "CureCode Installation Script..."

if not find first system/schemes 'MySQL [
	print "...loading MySQL protocol"
	do %mysql-protocol.r
]
print "...checking server status"
unless attempt [close open/no-wait/direct tcp://localhost:3306 true][
	print "*** Error: MySQL server unreachable on port 3306"
	halt
]

print "^/...creating database"
pass: ask/hide "...database root password (ENTER if none): "
install-pass: either empty? pass [""][join ":" pass]

db: open replace mysql://root$PASS$@localhost/mysql "$PASS$" install-pass
print "...access ok to MySQL server"

db-name: ask "^/...name of the new CureCode database (ENTER for default): "
if empty? trim db-name [db-name: "bugs"]

print "...creating database"
send-sql db replace/all {
    DROP DATABASE IF EXISTS $DB$;
    CREATE DATABASE IF NOT EXISTS $DB$;
    USE $DB$;
} "$DB$" db-name

print "...creating tables"
send-sql db read %build-db.sql
close db

msg: {
Please add the following webapp definition to your %httpd.cfg file in the domain
config block of your choice :

	webapp [
		virtual-root 	"/curecode/"
		root-dir 		$PATH$
		locales-dir 	%private/instances/default/locales/
		timeout 		00:30
		
		databases 		[
			bugs mysql://root$PASS$@localhost/$DB$
		]
		locals [
			name		"CureCode"
			instance	%default/
		]
	]
}
replace msg "$PATH$" mold first split-path what-dir
replace msg "$PASS$" install-pass
replace/all msg "$DB$" db-name

print msg

print {
Administrator account
---------------------
login: admin
pass:  nimda

Please change admin's password once logged (in Profile menu).
}

print "^/...intallation finished."
halt

private/access-control.r
REBOL []

comment {
	4 admin -> all
	3 developer -> !manager, !not-other-projects
	2 reporter -> !manager, !not-other-projects, !edit-other-ticket, !delete-ticket
	1 viewer -> !manager, !edit-ticket, !delete-ticket, !add-comment
}

deny-access: does [
	;response/redirect join request/web-app "/denied.html"
	response/redirect join request/web-app "/index.rsp"
]

filter-access: does [
	if not switch session/content/user-role [
		viewer [
			find [
				"index.rsp"
				"register.rsp"
				"view-tickets.rsp"
				"ticket.rsp"
				"validate.rsp"
				"set-project.rsp"
				"reset-pass.rsp"
				"reset.rsp"
				"captcha.rsp"
				"api.rsp"
				"attached.rsp"
				"get-file.rsp"
				"feed.rsp"
			] request/parsed/target
		]
		reporter [
			not find request/parsed/path "manage"	
		]
		developer [
			not find request/parsed/path "manage"	
		]
		admin [
			yes
		]
	][	
		deny-access
	]
]

private/generate-load.r
REBOL []

random/seed now/time/precise 

if not find first system/schemes 'MySQL [
	do %mysql-protocol.r
]

make-text: has [out][
	clear out: ""
	loop 32 [append out #"a" + ((random 26) - 1)]
	out
]
prin "Database root "
db: open mysql://root:?@localhost/bugs

repeat cnt 4 [
	send-sql db rejoin [ {
		INSERT INTO users VALUES (
			NULL,
			'dk} cnt {',
			'098F6BCD4621D373CADE4E832627B4F6',
			'default@curecode.org',
			'00000000000000000000000000000000',
			NOW(),
			2,
			NOW()
		)}
	]
	send-sql db rejoin [
		"INSERT INTO projects VALUES (NULL,'Test" cnt "','This is a test project',0,0)"
	]
	repeat ver 4 [
		send-sql db rejoin [
			"INSERT INTO versions VALUES (NULL," cnt + 1 ",'1." ver "',0)"
		]
	]
	send-sql db rejoin ["INSERT INTO categories VALUES (NULL," cnt + 1 ",'User Interface',0)"]
	send-sql db rejoin ["INSERT INTO categories VALUES (NULL," cnt + 1 ",'Security',0)"]
	send-sql db rejoin ["INSERT INTO categories VALUES (NULL," cnt + 1 ",'Text',0)"]
	send-sql db rejoin ["INSERT INTO categories VALUES (NULL," cnt + 1 ",'Performances',0)"]
]

counts: [0 0 0 0]

repeat cnt 10000 [
	either zero? cnt // 1000 [prin newline prin cnt][
		if zero? cnt // 100 [prin #"."]
	]
	prj: random 4
	poke counts prj counts/:prj + 1
	send-sql db rejoin [ 
		"INSERT INTO tickets VALUES (NULL,"
		prj + 1 ","
		"'" join make-text "',"
		"'" join make-text "',"
		"'" join make-text "',"
		1 + (prj * random 4) ","
		1 + (prj * random 4) ","
		1 + (prj * random 4) ","
		random 8 ","
		random 7 ","
		random 9 ","
		random 6 ","
		random 6 ","
		"0,"
		1 + random 4 ","	;avoid 'admin' user
		"NOW(), NOW())"
	]
]

repeat cnt 4 [
	send-sql db rejoin ["UPDATE projects SET count=" counts/:cnt " WHERE id=" cnt + 1]
]

close db

halt